compiler/rustc_trait_selection/src/traits/fulfill.rs

```diff
@@ -709,6 +712,17 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
             // no type variables present, can use evaluation for better caching.
             // FIXME: consider caching errors too.
             if self.selcx.infcx().predicate_must_hold_considering_regions(obligation) {
+                if let Some(key) = ProjectionCacheKey::from_poly_projection_predicate(
+                    &mut self.selcx,
+                    project_obligation.predicate,
+                ) {
+                    self.selcx
+                        .infcx()
+                        .inner
+                        .borrow_mut()
+                        .projection_cache()
+                        .complete(key, EvaluationResult::EvaluatedToOk);
```

Comment:
Currently we can `complete` a projection in two places. I'm curious which of the two gives the perf benefits. It could be that one is essentially useless (in which case that might suggest that code path is not followed much - at least in terms where caching could be helpful).

Either way, I think I'd at least like some better comments about *why* we can cache at each place.